File: internal/projects/branches.go
package projects

import (
	"path/filepath"
	"strings"
)

// NormalizePreferredBranchNames recorta, deduplica y deja como máximo 3 nombres.
func NormalizePreferredBranchNames(v []string) []string {
	seen := make(map[string]struct{})
	var out []string
	for _, s := range v {
		s = strings.TrimSpace(s)
		if s == "" {
			continue
		}
		if _, ok := seen[s]; ok {
			continue
		}
		seen[s] = struct{}{}
		out = append(out, s)
		if len(out) >= 3 {
			break
		}
	}
	return out
}

// NormalizePreferredBranchesMap normaliza claves a rutas absolutas limpias y los valores.
func NormalizePreferredBranchesMap(m map[string][]string) map[string][]string {
	if len(m) == 0 {
		return nil
	}
	out := make(map[string][]string)
	for k, v := range m {
		abs, err := filepath.Abs(k)
		if err != nil {
			continue
		}
		abs = filepath.Clean(abs)
		nv := NormalizePreferredBranchNames(v)
		if len(nv) == 0 {
			continue
		}
		out[abs] = nv
	}
	if len(out) == 0 {
		return nil
	}
	return out
}

// ApplyPreferredFirst devuelve all reordenado: primero las que aparecen en preferred (en ese orden),
// luego el resto en el orden original de all. Solo cuenta coincidencias exactas de nombre de ref.
func ApplyPreferredFirst(all []string, preferred []string) []string {
	if len(preferred) == 0 {
		out := make([]string, len(all))
		copy(out, all)
		return out
	}
	seen := make(map[string]struct{})
	var head []string
	for _, p := range preferred {
		p = strings.TrimSpace(p)
		if p == "" {
			continue
		}
		for _, a := range all {
			if a == p {
				if _, ok := seen[a]; !ok {
					head = append(head, a)
					seen[a] = struct{}{}
				}
				break
			}
		}
	}
	var rest []string
	for _, a := range all {
		if _, ok := seen[a]; !ok {
			rest = append(rest, a)
		}
	}
	return append(head, rest...)
}

File: ink-app/src/projects/branches.ts
import path from "node:path";

export function normalizePreferredBranchNames(v: string[]): string[] {
  const seen = new Set<string>();
  const out: string[] = [];
  for (const raw of v) {
    const s = raw.trim();
    if (!s || seen.has(s)) {
      continue;
    }
    seen.add(s);
    out.push(s);
    if (out.length >= 3) {
      break;
    }
  }
  return out;
}

export function normalizePreferredBranchesMap(
  m: Record<string, string[]> | undefined
): Record<string, string[]> {
  if (!m || Object.keys(m).length === 0) {
    return {};
  }
  const out: Record<string, string[]> = {};
  for (const [k, v] of Object.entries(m)) {
    const abs = path.resolve(k);
    const clean = path.normalize(abs);
    const nv = normalizePreferredBranchNames(v);
    if (nv.length === 0) {
      continue;
    }
    out[clean] = nv;
  }
  return out;
}

/** Preferred branch names first (exact ref match), then the rest in original order. */
export function applyPreferredFirst(
  all: string[],
  preferred: string[]
): string[] {
  if (preferred.length === 0) {
    return [...all];
  }
  const seen = new Set<string>();
  const head: string[] = [];
  for (const p of preferred) {
    const q = p.trim();
    if (!q) {
      continue;
    }
    for (const a of all) {
      if (a === q && !seen.has(a)) {
        head.push(a);
        seen.add(a);
        break;
      }
    }
  }
  const rest = all.filter((a) => !seen.has(a));
  return [...head, ...rest];
}
